File: 5. FEB 27/P1.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

struct Activity {
    int start, end;
};

bool compare(Activity a1, Activity a2) {
    return a1.end < a2.end;
}

void selectActivities(vector<Activity>& activities) {
    sort(activities.begin(), activities.end(), compare);

    cout << "Selected Activities (start, end):\n";
    int lastEnd = -1;
    for (auto act : activities) {
        if (act.start >= lastEnd) {
            cout << "(" << act.start << ", " << act.end << ")\n";
            lastEnd = act.end;
        }
    }
}

int main() {
    int n;
    cout << "Enter number of activities: ";
    cin >> n;

    vector<Activity> activities(n);
    cout << "Enter start and end time for each activity:\n";
    for (int i = 0; i < n; i++) {
        cin >> activities[i].start >> activities[i].end;
    }

    selectActivities(activities);
    return 0;
}

File: 5. FEB 27/PYTHON PROGRAMS/P1.py
import random
import time
import matplotlib.pyplot as plt

def activity_selection(activities):
    activities.sort(key=lambda x: x[1])
    count = 0
    last_end = -1
    for start, end in activities:
        if start >= last_end:
            count += 1
            last_end = end
    return count

input_sizes = [100 * i for i in range(1, 11)]
actual_times = []

for n in input_sizes:
    activities = [(random.randint(0, 50), random.randint(51, 100)) for _ in range(n)]
    start_time = time.time()
    activity_selection(activities)
    end_time = time.time()
    actual_times.append(end_time - start_time)

expected_times = [t * input_sizes[-1] / input_sizes[-1] * actual_times[-1] for t in input_sizes]

plt.plot(input_sizes, actual_times, color='teal', marker='o', label="Actual Time")
plt.plot(input_sizes, expected_times, 'r--', label="Expected O(n log n)")
plt.xlabel("Number of Activities")
plt.ylabel("Time (s)")
plt.title("Activity Selection Time Complexity")
plt.grid(True)
plt.legend()
plt.show()
